### PR TITLE
Common spec validation 

### PR DESCRIPTION
**Pull request by connorwstein**
- Assert a pipeline spec is present if required
- Pull top level checks into a helper
- Change modfile to go 1.16 as we now require it with the use of `io.NopCloser`
- Use the presenters for webhook job runs 
- Add txHash to the VRF duplicate request log

### CODE REVIEW

File: core/services/job/orm.go [hunk 1]
```diff
@@ -180,11 +180,6 @@ func (o *orm) CreateJob(ctx context.Context, jobSpec *Job, p pipeline.Pipeline)
 		}
 	}
 
-	if jobSpec.ExternalJobID == (uuid.UUID{}) {
```
Comment (connorwstein) [inline core/services/job/orm.go:183]: This is duplicated below, I think it was the result of a bad merge

File: core/services/offchainreporting/validate.go [hunk 2]
```diff
@@ -44,17 +44,10 @@ func ValidatedOracleSpecToml(config ValidationConfig, tomlString string) (job.Jo
 	}
 	jb.OffchainreportingOracleSpec = &spec
 
-	// TODO(#175801426): upstream a way to check for undecoded keys in go-toml
```
Comment (connorwstein) [inline core/services/offchainreporting/validate.go:47]: There is a way, see large comment in job/validate.go. Hinges on schemaVersion debate outcome

**Comment by jkongie** [comment]
Maybe we can consider moving the job spec parsing into a separate package which purely handles the validation and unmarshalling. Then in each job handling service we write a converter which converts the spec spec into the model that represents the business logic.

**Comment by connorwstein** [comment]
> Maybe we can consider moving the job spec parsing into a separate package which purely handles the validation and unmarshalling. Then in each job handling service we write a converter which converts the spec spec into the model that represents the business logic.

I did consider moving all the validation into job/validate.go but the OCR spec validation for example is pretty involved and depends on libocr types, so it seemed to nice to keep the job type specific validation in its respective packages. As for introducing an intermediate struct + converters seems like overkill at the moment - the toml specs essentially map 1-1 with the db models. Should there a be more complex mapping, then I'd be all for introducing an intermediate struct

File: core/services/fluxmonitorv2/validate_test.go [hunk 3]
```diff
@@ -129,32 +129,6 @@ ds1 -> ds1_parse;
 				c.Set("DEFAULT_HTTP_TIMEOUT", "2s")
 			},
 		},
-		{
-			name: "async=true should error",
```
Comment (samsondav) [inline core/services/fluxmonitorv2/validate_test.go:133]: Why has this gone?

File: core/services/job/models.go [hunk 4]
```diff
@@ -39,20 +39,33 @@ func (t Type) String() string {
...
 		FluxMonitor:       true,
-		OffchainReporting: true,
+		OffchainReporting: false, // bootstrap jobs do not require it
 		Keeper:            false,
 		VRF:               true,
 		Webhook:           true,
 	}
+	supportsAsync = map[Type]bool{
+		Cron:              false,
+		DirectRequest:     false,
```
Comment (samsondav) [inline core/services/job/models.go:62]: Pretty sure DirectRequest also supports async, no?

File: core/services/job/validate.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+// Common spec validation
+func ValidateSpec(ts string) (Type, error) {
+	var jb Job
+	// Note we can't use:
+	//   toml.NewDecoder(bytes.NewReader([]byte(ts))).Strict(true).Decode(&jb)
+	// to error in the case of unrecognized keys because all the keys in the toml are at
+	// the top level and so decoding for the job will have undecodable keys meant for the job
+	// type specific struct and vice versa. Should we upgrade the schema,
+	// we put the type specific config in its own subtree e.g.
+	// 	schemaVersion=1
```
Comment (samsondav) [inline core/services/job/validate.go:32]: This works, good solution

**Review by samsondav — APPROVED** [review]
Seems completely reasonable

File: core/services/job/models.go (hunk 4 above)
Comment (archseer) [reply core/services/job/models.go:62]: It doesn't yet, I'm working on that right now.

File: core/services/chainlink/application.go [hunk 6]
```diff
@@ -707,12 +707,8 @@ func (app *ChainlinkApplication) RunJobV2(
 	var runID int64
 
 	// Some jobs are special in that they do not have a task graph.
-	if !jb.Type.HasPipelineSpec() {
-		// This is a weird situation, even if a job doesn't have a pipeline it needs a pipeline_spec_id in order to insert the run
-		// TODO: Once all jobs have a pipeline this can be removed
-		// See: https://app.clubhouse.io/chainlinklabs/story/6065/hook-keeper-up-to-use-tasks-in-the-pipeline
-		runID, err = app.pipelineRunner.TestInsertFinishedRun(app.Store.DB.WithContext(ctx), jb.ID, jb.Name.String, jb.Type.String(), jb.PipelineSpecID)
-	} else {
+	isBootstrap := jb.Type == job.OffchainReporting && jb.OffchainreportingOracleSpec != nil && jb.OffchainreportingOracleSpec.IsBootstrapPeer
```
Comment (PiotrTrzpil) [inline core/services/chainlink/application.go:710]: might be worth creating a `IsBootstrap()` method in the job?

File: core/services/fluxmonitorv2/validate_test.go (hunk 3 above)
Comment (connorwstein) [reply core/services/fluxmonitorv2/validate_test.go:133]: validation is just done at a higher level now

File: core/services/chainlink/application.go (hunk 6 above)
Comment (connorwstein) [reply core/services/chainlink/application.go:710]: agreed, but probably best to just wait for https://app.clubhouse.io/chainlinklabs/story/13879/move-bootstrap-ocr-jobs-into-their-own-job-type

**Pull request merged**